Access PathsAndConsumesOfModules from new signal

wddgit opened the pull request:
#### PR description:

Use the new LookupInitializationComplete signal to get access to the PathsAndConsumesOfModules object that is available to Services. Previously the PreBeginJob signal was used for this purpose, but since PR #46929 was merged, the object is only partially filled at preBeginJob. When the new signal is emitted it is fully filled. PathsAndConsumesOfModules now includes information related to EventSetup modules that previously did not exist. That is the part that is not filled at preBeginJob.

This shouldn't affect the output of anything (except DependencyGraph service used with SubProcesses, see below).

This PR removes the argument from the preBeginJob signal interface. For several of the services modified in this PR, that argument is not used and the PR just deletes it from the argument list. These changes are trivial and the fact that those services compile should be a sufficient test.

This PR modifies other affected services to use the new signal.

- A few services are used in Framework unit tests and those test pass with the changes in this PR.
- DependencyGraph runs and produces a graph successfully. If one modifies it to run without SubProcesses, the output .gv file is identical before and after this PR. (Note that with SubProcesses the version of DependencyGraph before this PR has a bug that is fixed with the changes in this PR. The old version assumes the PostBeginRun signal is a "local" signal" in ActivityRegistry but it is "global" so none of the SubProcess info is included in the .gv file that is output in the old version.)
- FastTimerService is moved to the new signal and it looks like nothing should change in the output. There are many references to FastTimerService in CMSSW, but I don't see any unit test devoted to testing it. Is there an existing test of FastTimerService somewhere that I could run? Or is there an expert willing to test this? It is probably just a matter of running and checking the output is reasonable. Nothing should change.

There is one Service where the changes are more complex. That is the NVProfilerService. The PathsAndConsumesOfModules object is used to size some vectors at PreBeginJob. Not only is this case more complicated but the current implementation appears to be incorrect. There is a bug that could cause out of bounds vector access. The size of the ProcessCallGraph was being used to size the vectors, but that size could be incorrect if modules are deleted. The framework will delete unused modules. The ID from the ModuleDescription is incremented for every module constructed, including ones that are later deleted. This ID is not modified when a module is deleted. The index used to access the vector is the ID from the ModuleDescription, but in the old version the size is a count of undeleted modules from PathsAndConsumesOfModules and that does have deleted modules removed. This is fixed in the new version where the size is derived from the number of modules constructed and it counts the ones that later get deleted.

Note that there are no tests in the CMSSW repository of NVProfilerService. In fact there are no references to it at all in CMSSW. Is there a test of NVProfilerService somewhere that I could run? Or is there an expert willing to test this?

Should I develop tests for these two services? If necessary, this could be accomplished in a separate PR...

Resolves https://github.com/cms-sw/framework-team/issues/1194

#### PR validation:

Relied on existing unit tests which all pass.

wddgit committed:
Access PathsAndConsumesOfModules from new signal

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"c5a164e61f4707e62ced6d327aa3667658822d55":{"files":["EventFilter/Utilities/interface/EvFDaqDirector.h","EventFilter/Utilities/interface/FastMonitoringService.h","EventFilter/Utilities/src/FastMonitoringService.cc","FWCore/Framework/interface/GlobalSchedule.h","FWCore/Framework/interface/Schedule.h","FWCore/Framework/src/EventProcessor.cc","FWCore/Framework/src/GlobalSchedule.cc","FWCore/Framework/src/Schedule.cc","FWCore/Framework/src/SubProcess.cc","FWCore/Integration/plugins/PathsAndConsumesOfModulesTestService.cc","FWCore/Integration/plugins/TestServiceOne.cc","FWCore/Integration/plugins/TestServiceOne.h","FWCore/PrescaleService/interface/PrescaleService.h","FWCore/PrescaleService/src/PrescaleService.cc","FWCore/ServiceRegistry/interface/ActivityRegistry.h","FWCore/Services/plugins/CheckTransitions.cc","FWCore/Services/plugins/CondorStatusUpdater.cc","FWCore/Services/plugins/DependencyGraph.cc","FWCore/Services/plugins/Timing.cc","FWCore/Services/plugins/Tracer.cc","FWCore/Services/plugins/tracer_setupFile.cc","FWCore/TestProcessor/src/TestProcessor.cc","HLTrigger/Timer/interface/ProcessCallGraph.h","HLTrigger/Timer/plugins/FastTimerService.cc","HLTrigger/Timer/plugins/FastTimerService.h","HLTrigger/Timer/src/ProcessCallGraph.cc","HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc","HeterogeneousCore/SonicTriton/interface/TritonService.h","HeterogeneousCore/SonicTriton/src/TritonService.cc","IOMC/RandomEngine/plugins/RandomNumberGeneratorService.cc","IOMC/RandomEngine/plugins/RandomNumberGeneratorService.h","PerfTools/AllocMonitor/plugins/ModuleEventAllocMonitor.cc","PerfTools/AllocMonitor/plugins/moduleAlloc_setupFile.cc"],"squashed":false,"time":1740513068}},"emoji":{},"last_seen_sha":"c5a164e61f4707e62ced6d327aa3667658822d55","signatures":{}} -->

cmsbuild commented:
A new Pull Request was created by  @wddgit for master.

It involves the following packages:

- EventFilter/Utilities (**daq**)
- FWCore/Framework (**core**)
- FWCore/Integration (**core**)
- FWCore/PrescaleService (**core**)
- FWCore/ServiceRegistry (**core**)
- FWCore/Services (**core**)
- FWCore/TestProcessor (**core**)
- HLTrigger/Timer (**hlt**)
- HeterogeneousCore/CUDAServices (**heterogeneous**)
- HeterogeneousCore/SonicTriton (**heterogeneous**)
- IOMC/RandomEngine (**core**)
- PerfTools/AllocMonitor (**core**)


@Dr15Jones, @Martin-Grunewald, @cmsbuild, @emeschi, @fwyzard, @makortel, @mmusich, @smorovic, @smuzaffar can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @fabiocos, @felicepantaleo, @fwyzard, @kpedro88, @makortel, @missirol, @mmusich, @riga, @rovere this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

wddgit commented:
I just noticed that FastTimerService will also see a behavior change caused by this PR if used with SubProcesses. The activity that used to be in PostBeginJob would not have been executed for SubProcess signals and now would be. Not sure if this matters as we are not currently using SubProcess and discussing removing the SubProcess feature entirely. Possibly this fixes a problem also.

The behavior of other services would not be affected by this PR, but they might be having similar problems if used with SubProcess. NVProfilerService uses the PostBeginJob signal and might or might not be affected (even if it has nothing to do with this PR).

makortel commented:
> Should I develop tests for these two services?

No.

mmusich commented:
test parameters:
   * enable =  hlt_p2_timing

mmusich commented:
@cmsbuild, please test
   * https://github.com/cms-sw/cmssw/pull/47467#issuecomment-2685876789 should guarantee some form of the `FastTimerService` is run (and we can check the impact).

wddgit commented:
@mmusich Thanks.

fwyzard commented:
hold

fwyzard commented:
until I will have time to assess the impact on the FastTimerService

cmsbuild commented:
Pull request has been put on hold by  @fwyzard
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

wddgit commented:
I was looking back at the comments above and they seem to be confusing, so I thought I would provide some more background information related to SubProcess.

First, I want to emphasize this only applies to jobs including one or more SubProcesses. None of this affects jobs without a SubProcess.

None of what I am explaining in this comment is new. The code has behaved like this for more than a decade.

A signal emitted in a job with SubProcesses can be propagated in two ways. "global" signals are propagated to services in the process or SubProcess where it was emitted and to child SubProcesses. So a "global" signal emitted in the top level process would be seen by services in the top level processes and all SubProcesses. But a "global" signal emitted in a SubProcess would not be seen by a service in the top level process or any higher level SubProcess.

A "local" signal goes in the opposite direction. A service in the top level process will see every "local" signal emitted in the top level process or any SubProcess. The signal is propagated upward from the SubProcess it is emitted in.

This is the place where signals are assigned as global or local:

https://cmssdt.cern.ch/lxr/source/FWCore/ServiceRegistry/src/ActivityRegistry.cc#0056

And finally we are discussing deleting the SubProcess feature entirely and currently no one uses it. So maybe I'm wasting time discussing this...

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 212KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b40efd/44702/summary.html
**COMMIT**: c5a164e61f4707e62ced6d327aa3667658822d55
**CMSSW**: CMSSW_15_1_X_2025-02-26-1100/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47467/44702/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test Zmumuall had ERRORS
</pre>

* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_1_X_2025-02-26-1100&dataset=CMSSW_15_1_X_2025-02-26-1100-cmssw47467/el8_amd64_gcc12/86/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - You potentially [added 6 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-26-1100+b40efd/67171/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 11 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-26-1100+b40efd/67171/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3920300
 - DQMHistoTests: [Total failures: 11](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-26-1100+b40efd/67171/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3920269
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-26-1100+b40efd/67171/triggerResults): no differences found

makortel commented:
The unit test failure is
```
===== Test "Zmumuall" ====
file has vanished: "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_15_1_X_2025-02-26-1100/unit_tests/validateAlignments/test_yaml/JetHT/merge/testJob/unitTestJetHT/JetHTAnalysis_merged.root"
rsync warning: some files vanished before they could be transferred (code 24) at main.c(1189) [sender=3.1.3]

---> test Zmumuall had ERRORS
```
Not sure what that means

fwyzard commented:
> And finally we are discussing deleting the SubProcess feature entirely and currently no one uses it. So maybe I'm wasting time discussing this...

If subprocesses are removed from CMSSW, I the FastTimerService can be cleaned up, and these changes may become irrelevant.
As long as subprocesses are part of CMSSW, I want to make sure that the FastTimerService works well with them.

wddgit commented:
The postBeginJob signal has been a "global" signal since January 2011 when SubProcess was first implemented by Bill. Chris and I were around then and also others when those commits were merged and involved in the discussions. If a postBeginJob signal is emitted in a SubProcess, then a service in the top level process does not see it and the postBeginJob method is not called. Any services relying on that signal coming from SubProcesses were already broken before this PR (only for jobs including a SubProcess). They have probably always been broken. I have not surveyed all services to determine which services suffer from this problem. The commit is back on the old CVS branch:

https://github.com/cms-sw/cmssw/commit/f4d9ee05d837830efdf17b6f31b9ec79f9b5dc95#diff-8138d6b90e119ac25a32ffb17e3b852c5c8f2588b8eb0f7300a0097fedf4e80c

This PR moves work from the postBeginJob signal to the new LookupInitializationComplete signal for the 3 services, DependencyGraph, FastTimerService, and NVProfilerService. The new signal is "local" so the functions will start being executed with a SubProcess in the job. I have not tested this, but this might be enough to fix them. Although it is probably safe to assume they were never tested with SubProcesses before so it is difficult to say with certainty that this will be enough to fix them. Maybe more needs to be done. (I verified this behavior with DependencyGraph, I didn't test FastTimerService or NVProfilerService with a SubProcess yet.)

I'm happy to fix something if it can be identified as wrong and be quickly fixed. I'm not sure it makes sense to hold up this PR for a long time for something that is already broken. We also could follow up with additional PRs to get specific services working with SubProcess if anyone has time to do that.

I'll take a look at the Zmumuall unit test failure. I was hoping someone else would comment about that. At a glance, I've no idea what that one is related to.

wddgit commented:
please test

Try it again. See if the unit test is reproducible.

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b40efd/44728/summary.html
**COMMIT**: c5a164e61f4707e62ced6d327aa3667658822d55
**CMSSW**: CMSSW_15_1_X_2025-02-27-1100/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47467/44728/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_1_X_2025-02-27-1100&dataset=CMSSW_15_1_X_2025-02-27-1100-cmssw47467/el8_amd64_gcc12/87/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-27-1100+b40efd/67196/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3920300
 - DQMHistoTests: [Total failures: 57](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-27-1100+b40efd/67196/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3920223
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-27-1100+b40efd/67196/triggerResults): no differences found

wddgit commented:
Tests pass this time.  Comparison failures are ones we've seen before in other PRs:

- workflow 29834.999 see Issue #47071
- in MessageLogger
- in LogErrorHarvester

The unit test failure we saw in the first run of the PR tests vanished. It appears unrelated to this PR to me. I cannot see how the changes in this PR in services that are mostly monitoring things could cause this error. Given that the error doesn't reproduce, I'm inclined to write it off as a testing system glitch, but if anyone has suggestions on anything I could do related to it I am all ears.

```
===== Test "Zmumuall" ====
file has vanished: "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_15_1_X_2025-02-26-1100/unit_tests/validateAlignments/test_yaml/JetHT/merge/testJob/unitTestJetHT/JetHTAnalysis_merged.root"
rsync warning: some files vanished before they could be transferred (code 24) at main.c(1189) [sender=3.1.3]

---> test Zmumuall had ERRORS
```

wddgit commented:
I see nice looking HLT circle charts by following the link above but am not familiar with them enough to say whether they are good or bad. I do not know whether they are populated by output of the FastTimerService or not.

mmusich commented:
> I see nice looking HLT circle charts by following the link above but am not familiar with them enough to say whether they are good or bad. I do not know whether they are populated by output of the FastTimerService or not.

the input data used to populate those diagrams is fetched via `FastTimerService`. The test is not meant to be exhaustive and it's actually meant for other reasons. The fact that you see circles filled it means that nothing is *badly* broken in a way that would make that particular test unusable. 
There are subtleties though (see Andrea's comment https://github.com/cms-sw/cmssw/pull/47467#issuecomment-2690015350 to which I completely defer as main developer and expert of the tool).

wddgit commented:
@fwyzard Thanks for the comments. I took a closer look.

Before I continue, I want to emphasize all of this only affects jobs using SubProcess. As far as I know, no one is currently using SubProcess.

Two apologies. First, I see that FastTimerService did work with SubProcess in the past. I was wrong when I said it didn't. Second, I am the one that broke it in July 2024. It was an unintentional side effect of PR #45434. My mistake. That PR was one of a series of PRs where we went through all the begin and end transitions and tried to give them consistent behavior, especially under exceptions. Prior to that PR each different pair of begin/end transitions behaved differently and it took a careful look at code to understand how a particular begin/end pair of transitions behaved.

So since July 2024, several services have been broken when using SubProcesses, including FastTimerService. As far as I know I am the first to notice. I noticed while testing this PR. Our PR and IB tests aren't good enough to detect the problem.

There are 2 ways forward:

- Delete SubProcess and accept that jobs with SubProcess are broken temporarily until we do that. If we are doing that soon, this is probably the best approach.
- I can revert the part of PR #45434 that is causing the problem. I would suggest doing that in a separate PR first. It also affects postEndJob, not just postBeginJob. And then I would also need to make a few changes to this PR. This isn't hard and we could still delete SubProcess in the future. Either way is OK with me.

makortel commented:
> * Delete SubProcess and accept that jobs with SubProcess are broken temporarily until we do that. If we are doing that soon, this is probably the best approach.

On a quick thought I'd be in favor just proceeding this way, but let's discuss on Monday.

wddgit commented:
An update. I am starting to work on a PR removing the SubProcess feature from the CMS Python code used to support configurations and also remove any SubProcess tests. That will completely disable it. Then subsequent PRs will follow that remove all traces of SubProcess from Core code and possibly CMSSW. Even before we realized there was a problem introduced last summer, Matti had proposed this in his O&C talk. I think he plans to announce it again, but if no one in CMS comes forward to say that they use SubProcess or need it for the future for some specific reason, it seems to be where we are headed.

fwyzard commented:
@wddgit thanks for looking into this and for the additional details.


> * Delete SubProcess and accept that jobs with SubProcess are broken temporarily until we do that. If we are doing that soon, this is probably the best approach.

OK, I am not going to be the one asking to keep support for subprocess in CMSSW if nobody is using it.

makortel commented:
> I think he plans to announce it again, but if no one in CMS comes forward to say that they use SubProcess or need it for the future for some specific reason, it seems to be where we are headed.

Now announced in https://cms-talk.web.cern.ch/t/removal-of-subprocess-feature/117449

wddgit commented:
PR #47565 removed support for SubProcess from the Framework configuration system and has been merged into the master branch. SubProcess is now permanently disabled and it is now impossible to run a process with a SubProcess. There will be PRs in the future to cleanup the remnants of SubProcess from other parts of the Framework code, but that is just cleanup.

It's my belief that with that accomplished, this PR is now ready to be merged without modification. It accomplishes its stated purpose. If any one has further comments please let me know and I'll address those.

FastTimerService, DependenceGraph and ProcessCallGraph will need further modifications to remove code in them related to SubProcess. And maybe also NVProfilerService, someone should look at that more carefully. At least to me, it would seem to be better to handle that in a separate dedicated PR. @makortel @fwyzard I would be happy to work on that . Also happy to move on to other things if the maintainers of those services would prefer to take care of this. Just let me know which you guys prefer. I could also implement it and someone else could test it. I'm going to look again more closely, but so far those are the only things I've noticed outside of Core Framework code that depend on SubProcess code.

makortel commented:
@cmsbuild, please test

makortel commented:
> At least to me, it would seem to be better to handle that in a separate dedicated PR.

I'm in favor of doing the SubProcess cleanup in subsequent PR(s), and not hold this PR for them.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b40efd/45007/summary.html
**COMMIT**: c5a164e61f4707e62ced6d327aa3667658822d55
**CMSSW**: CMSSW_15_1_X_2025-03-16-2300/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47467/45007/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 5 errors in the following unit tests:

<pre>---> test testTauEmbeddingWorkflow2016preVFP had ERRORS
---> test testTauEmbeddingWorkflow2017 had ERRORS
---> test testTauEmbeddingWorkflow2016postVFP had ERRORS
and more ...
</pre>

* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_1_X_2025-03-16-2300&dataset=CMSSW_15_1_X_2025-03-16-2300-cmssw47467/el8_amd64_gcc12/91/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - You potentially [added 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-16-2300+b40efd/67452/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-16-2300+b40efd/67452/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3927607
 - DQMHistoTests: [Total failures: 7](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-16-2300+b40efd/67452/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3927580
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-16-2300+b40efd/67452/triggerResults): no differences found

makortel commented:
The test failures occur in IBs and are related to https://github.com/cms-sw/cmssw/pull/47572#issuecomment-2725429768

makortel commented:
+core

makortel commented:
ignore tests-rejected with ib-failure

fwyzard commented:
I'm OK to clean up the Subprocess remaining code in a follow up PR.

fwyzard commented:
unhold

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b40efd/45119/summary.html
**COMMIT**: c5a164e61f4707e62ced6d327aa3667658822d55
**CMSSW**: CMSSW_15_1_X_2025-03-20-1100/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47467/45119/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_1_X_2025-03-20-1100&dataset=CMSSW_15_1_X_2025-03-20-1100-cmssw47467/el8_amd64_gcc12/93/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 6 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-20-1100+b40efd/67579/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3929449
 - DQMHistoTests: [Total failures: 119](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-20-1100+b40efd/67579/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3929310
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-20-1100+b40efd/67579/triggerResults): no differences found

wddgit commented:
This PR is still waiting for review from daq, heterogeneous, and hlt. If you have comments, please let me know and I will try to  resolve them, either answer questions or fix things...

smorovic commented:
+daq

fwyzard commented:
> * FastTimerService is moved to the new signal and it looks like nothing should change in the output. There are many references to FastTimerService in CMSSW, but I don't see any unit test devoted to testing it. Is there an existing test of FastTimerService somewhere that I could run? Or is there an expert willing to test this? It is probably just a matter of running and checking the output is reasonable. Nothing should change.

I will double check the `FastTimerService`.


> There is one Service where the changes are more complex. That is the NVProfilerService. The PathsAndConsumesOfModules object is used to size some vectors at PreBeginJob. Not only is this case more complicated but the current implementation appears to be incorrect. There is a bug that could cause out of bounds vector access. The size of the ProcessCallGraph was being used to size the vectors, but that size could be incorrect if modules are deleted. 

I see.
The `NVProfilerService` definitely predates the ability of the framework to delete unused modules, so it's possible that it wasn't updated when the ability was introduced.

I will have a look at the `NVProfilerService` as well.

fwyzard commented on HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc:
```diff
@@ -543,6 +541,11 @@ void NVProfilerService::postBeginJob() {
 
+void NVProfilerService::lookupInitializationComplete(edm::PathsAndConsumesOfModulesBase const& pathsAndConsumes,
+                                                     edm::ProcessContext const& context) {
+  callgraph_.lookupInitializationComplete(pathsAndConsumes, context);
```
@wddgit could you add a marker here ?
```c++
  nvtxDomainMark(global_domain_, "lookupInitializationComplete");
```

Or, if you think that displaying some information from the `lookupInitializationComplete` signal could be interesting, something like
```c++
  std::stringstream out;
  out << "lookupInitializationComplete: " << ...;
  nvtxDomainMark(global_domain_, out.str().c_str());
```

wddgit replied:
Done. I added the call to nvtxDomainMark.

fwyzard commented on HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc:
```diff
@@ -831,7 +843,6 @@ void NVProfilerService::postModuleConstruction(edm::ModuleDescription const& des
 void NVProfilerService::preModuleDestruction(edm::ModuleDescription const& desc) {
   if (not skipFirstEvent_) {
     auto mid = desc.id();
-    global_modules_.grow_to_at_least(mid + 1);
```
@wddgit this is no longer needed because `global_modules_` will have already been grown in `preModuleConstruction` ?

wddgit replied:
Yes.

fwyzard commented on HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc:
```diff
@@ -514,6 +518,10 @@ void NVProfilerService::preallocate(edm::service::SystemBounds const& bounds) {
   stream_modules_.resize(concurrentStreams);
+  for (auto& modulesForOneStream : stream_modules_) {
+    modulesForOneStream.resize(global_modules_.size(), nvtxInvalidRangeId);
+  }
```
@wddgit does this mean that, by the time `preallocate()` is called, all modules have already been constructed, and `global_modules_` has its final size ?

wddgit replied:
Yes. But there is one exception to that. In the case of a looper, it is possible for the looper to replace existing modules in loops after the first one with newly constructed modules in the end of loop transition.

I'm not sure if anyone actually uses that feature. In the CMSSW repository, I can see a Framework unit test and Fireworks with calls into the module replacement interface...

fwyzard commented on HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc:
```diff
@@ -811,9 +814,18 @@ void NVProfilerService::postModuleEventPrefetching(edm::StreamContext const& sc,
+  // called. But there is a rare case when a looper is used that replacement
+  // modules can be constructed at end of loop. I'm not sure if that feature
+  // is ever actually used but just to be safe...
+  for (auto& modulesForOneStream : stream_modules_) {
```
@wddgit note that `stream_modules_` is a regular `std::vector<std::vector<...>>`.

Can multiple `preModuleConstruction` happen in parallel in different threads ?

If so, and for any reason `stream_modules_` is not empty, this will try to resize the same `vector` concurrently, which is not thread safe.

wddgit replied:
In the present version of the Framework, modules are constructed serially. There is not any concurrency. We've sometimes talked about changing that, but it would be difficult and would cause a lot of difficult to find and fix problems here and elsewhere.

stream_modules will always be empty in preModuleConstruction except in the case where it is called during end of loop via the looper. The end of loop code that replaces modules is serial.

And one other thing related that might be worth mentioning is that the source constructor is called concurrently with module construction normally. Sources are never constructed at end of loop so that is not an issue. Also, according to the TBB documentation grow_to_at_least can be called concurrently.

fwyzard replied:
Indeed, I'm not concerned about `global_modules_`: that is a `tbb::concurrent_vector` specifically to support growing in parallel, in `preSourceConstruction` and in `preModuleConstruction`.

The only concern I have is with the `modulesForOneStream.resize(...)`.

If the `preModuleConstruction` will always be called one module at a time, then it should be fine.

fwyzard commented on HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc:
```diff
@@ -523,18 +531,8 @@ void NVProfilerService::preallocate(edm::service::SystemBounds const& bounds) {
-
-  // this assumes that preBeginJob is not called concurrently with the modules' beginJob method
-  // or the preBeginJob for a subprocess
-  unsigned int modules = callgraph_.size();
```
This was the only use of the `callgraph_`.
If it is no longer needed to get the correct number of modules (because it is already known in `preallocate()`), the `callgraph_` object can be removed.

wddgit replied:
I removed callgraph_. That's better. Thanks.

fwyzard commented:
@wddgit sorry it took me so long to get back to this.
I'm not sure I understand all the changes related to the module ids, but I guess I should test the changes directly to follow better.
Anyway, I've left a few comments about the `NVProfilerService`.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47467/44475

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/Schedule.h  modified in PR(s): #47659
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #47659
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #47659
  - File  FWCore/Framework/src/SubProcess.cc  modified in PR(s): #47659
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #47659
  - File  HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc  modified in PR(s): #47830

cmsbuild commented:
Pull request #47467 was updated. @Dr15Jones, @Martin-Grunewald, @cmsbuild, @emeschi, @fwyzard, @makortel, @mmusich, @smorovic, @smuzaffar can you please check and sign again.

wddgit commented:
please test

fwyzard commented:
@wddgit thanks for the update. The only question I still have is [here](https://github.com/cms-sw/cmssw/pull/47467#pullrequestreview-2754165586) .

wddgit commented:
I think I resolved the comments. Let me know if there are more. Two changes since you looked at it:

- Removed callgraph_
- Added call to nvtxDomainMark

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b40efd/45503/summary.html
**COMMIT**: 634b7fd5beadb031180dfb78311a9f9c58af3e2b
**CMSSW**: CMSSW_15_1_X_2025-04-10-1100/el8_amd64_gcc12
**Additional Tests**: HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47467/45503/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_1_X_2025-04-10-1100&dataset=CMSSW_15_1_X_2025-04-10-1100-cmssw47467/el8_amd64_gcc12/95/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - You potentially [added 4 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-10-1100+b40efd/67942/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 8 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-10-1100+b40efd/67942/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3916361
 - DQMHistoTests: [Total failures: 14](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-10-1100+b40efd/67942/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3916327
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-10-1100+b40efd/67942/triggerResults): no differences found

makortel commented:
+core

fwyzard commented:
+heterogeneous

wddgit commented:
This might conflict with #47659. It will be trivial to fix, but I will rebase on top of that one as soon as it is merged. (Possibly git will be smart enough to handle it automatically). Probably not a good idea to merge them at the same time...

mmusich commented:
+hlt

makortel commented:
Since this PR has more signatures than https://github.com/cms-sw/cmssw/pull/47659, maybe it would be better to merge this one first if @cms-sw/daq-l2 signs soon?

wddgit commented:
Either order should work fine. Just not both at the same time. I agree that doing this one first makes sense if we get all the signatures.

smorovic commented:
+daq

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @mandrenguyen, @antoniovilela, @sextonkennedy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
